// @sf/functions/lib/commands/env/logdrain/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const color_1 = require("@heroku-cli/color");
const core_1 = require("@oclif/core");
const cli_ux_1 = require("cli-ux");
const core_2 = require("@salesforce/core");
const flags_1 = require("../../../lib/flags");
const base_1 = require("../../../lib/base");
const utils_1 = require("../../../lib/utils");
core_2.Messages.importMessagesDirectory(__dirname);
const messages = core_2.Messages.loadMessages('@salesforce/plugin-functions', 'env.logdrain.add');
class LogDrainAdd extends base_1.default {
    async run() {
        var _a, _b;
        const { flags } = await this.parse(LogDrainAdd);
        // We support both versions of the flag here for the sake of backward compat
        const targetCompute = (_a = flags['target-compute']) !== null && _a !== void 0 ? _a : flags.environment;
        const url = (_b = flags['drain-url']) !== null && _b !== void 0 ? _b : flags.url;
        if (!targetCompute) {
            throw new core_1.Errors.CLIError(`Missing required flag:
        -c, --target-compute TARGET-COMPUTE  ${color_1.default.dim('Environment name.')}
       See more help with --help`);
        }
        if (!url) {
            throw new core_1.Errors.CLIError(`Missing required flag:
       -u, --drain-url DRAIN-URL  ${color_1.default.dim('Endpoint that will receive sent logs.')}
       See more help with --help`);
        }
        if (flags.environment) {
            this.warn(messages.getMessage('flags.environment.deprecation'));
        }
        if (flags.url) {
            this.warn(messages.getMessage('flags.url.deprecation'));
        }
        const appName = await (0, utils_1.resolveAppNameForEnvironment)(targetCompute);
        cli_ux_1.cli.action.start(`Creating drain for environment ${color_1.default.app(targetCompute)}`);
        await this.client.post(`/apps/${appName}/log-drains`, {
            data: {
                url,
            },
        });
        cli_ux_1.cli.action.stop();
    }
}
exports.default = LogDrainAdd;
LogDrainAdd.summary = messages.getMessage('summary');
LogDrainAdd.description = messages.getMessage('description');
LogDrainAdd.examples = messages.getMessages('examples');
LogDrainAdd.flags = {
    'target-compute': flags_1.FunctionsFlagBuilder.environment({
        exclusive: ['environment'],
    }),
    environment: flags_1.FunctionsFlagBuilder.environment({
        char: 'e',
        exclusive: ['target-compute'],
        hidden: true,
    }),
    'drain-url': core_1.Flags.string({
        exclusive: ['url'],
        char: 'l',
        description: messages.getMessage('flags.drain-url.summary'),
    }),
    url: core_1.Flags.string({
        exclusive: ['drain-url'],
        char: 'u',
        description: messages.getMessage('flags.drain-url.summary'),
        hidden: true,
    }),
};
