// @sf/functions/lib/commands/env/var/unset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const color_1 = require("@heroku-cli/color");
const core_1 = require("@oclif/core");
const cli_ux_1 = require("cli-ux");
const core_2 = require("@salesforce/core");
const flags_1 = require("../../../lib/flags");
const utils_1 = require("../../../lib/utils");
const base_1 = require("../../../lib/base");
core_2.Messages.importMessagesDirectory(__dirname);
const messages = core_2.Messages.loadMessages('@salesforce/plugin-functions', 'env.var.unset');
class ConfigUnset extends base_1.default {
    async run() {
        var _a;
        const { flags, argv } = await this.parse(ConfigUnset);
        // We support both versions of the flag here for the sake of backward compat
        const targetCompute = (_a = flags['target-compute']) !== null && _a !== void 0 ? _a : flags.environment;
        if (!targetCompute) {
            throw new core_1.Errors.CLIError(`Missing required flag:
        -c, --target-compute TARGET-COMPUTE  ${color_1.default.dim('Environment name.')}
       See more help with --help`);
        }
        if (flags.environment) {
            this.warn(messages.getMessage('flags.environment.deprecation'));
        }
        const appName = await (0, utils_1.resolveAppNameForEnvironment)(targetCompute);
        if (argv.length === 0) {
            throw new core_1.Errors.CLIError('you must enter a config var key (i.e. mykey)');
        }
        const configPairs = argv.reduce((acc, elem) => {
            return Object.assign(Object.assign({}, acc), { [elem]: null });
        }, {});
        cli_ux_1.cli.action.start(`Unsetting ${Object.keys(configPairs)
            .map((key) => color_1.default.configVar(key))
            .join(', ')} and restarting ${color_1.default.app(targetCompute)}`);
        await this.client.patch(`/apps/${appName}/config-vars`, {
            data: configPairs,
        });
        cli_ux_1.cli.action.stop();
    }
}
exports.default = ConfigUnset;
ConfigUnset.strict = false;
ConfigUnset.summary = messages.getMessage('summary');
ConfigUnset.description = messages.getMessage('description');
ConfigUnset.examples = messages.getMessages('examples');
ConfigUnset.flags = {
    'target-compute': flags_1.FunctionsFlagBuilder.environment({
        exclusive: ['environment'],
    }),
    environment: flags_1.FunctionsFlagBuilder.environment({
        char: 'e',
        exclusive: ['target-compute'],
        hidden: true,
    }),
};
