// @salesforce/sf-plugins-core/lib/flags/orgFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requiredHubFlag = exports.requiredOrgFlag = exports.optionalOrgFlag = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const core_1 = require("@oclif/core");
const core_2 = require("@salesforce/core");
core_2.Messages.importMessagesDirectory(__dirname);
const messages = core_2.Messages.loadMessages('@salesforce/sf-plugins-core', 'messages');
const maybeGetOrg = async (input) => core_2.Org.create({ aliasOrUsername: input });
const getOrgOrThrow = async (input) => {
    const org = await maybeGetOrg(input);
    if (!org) {
        throw messages.createError('errors.NoDefaultEnv');
    }
    return org;
};
const getHubOrThrow = async (aliasOrUsername) => {
    var _a, _b, _c;
    if (!aliasOrUsername) {
        // check config for a default
        const config = await core_2.ConfigAggregator.create();
        aliasOrUsername =
            (_b = (_a = config.getInfo('target-dev-hub')) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : (_c = config.getInfo('defaultdevhubusername')) === null || _c === void 0 ? void 0 : _c.value;
        if (!aliasOrUsername) {
            throw messages.createError('errors.NoDefaultDevHub');
        }
    }
    const org = await core_2.Org.create({ aliasOrUsername });
    // check the synchronous, cached version and only use "determine" if we didn't get a positive result
    if (!org.isDevHubOrg() || !(await org.determineIfDevHubOrg())) {
        throw messages.createError('errors.NotADevHub', [aliasOrUsername]);
    }
    return org;
};
/**
 * An optional org specified by username or alias
 * Will default to the default org if one is not specified.
 * Will not throw if the specified org and default do not exist
 *
 * @example
 * import { SfCommand, Flags } from '@salesforce/sf-plugins-core';
 * public static flags = {
 *     // setting length or prefix
 *    'target-org': Flags.optionalOrg(),
 *    // adding properties
 *    'flag2': Flags.optionalOrg({
 *        required: true,
 *        description: 'flag2 description',
 *     }),
 * }
 */
exports.optionalOrgFlag = core_1.Flags.build({
    char: 'e',
    parse: async (input) => await maybeGetOrg(input),
    default: async () => await maybeGetOrg(),
    defaultHelp: async () => { var _a; return (_a = (await maybeGetOrg(undefined))) === null || _a === void 0 ? void 0 : _a.getUsername(); },
});
/**
 * An required org, specified by username or alias
 * Will throw if the specified org default do not exist
 * Will default to the default org if one is not specified.
 * Will throw if no default org exists and none is specified
 *
 * @example
 * import { SfCommand, Flags } from '@salesforce/sf-plugins-core';
 * public static flags = {
 *     // setting length or prefix
 *    'target-org': Flags.requiredOrg(),
 *    // adding properties
 *    'flag2': Flags.requiredOrg({
 *        required: true,
 *        description: 'flag2 description',
 *        char: 'o'
 *     }),
 * }
 */
exports.requiredOrgFlag = core_1.Flags.build({
    char: 'e',
    parse: async (input) => await getOrgOrThrow(input),
    default: async () => await getOrgOrThrow(undefined),
    defaultHelp: async () => { var _a; return (_a = (await getOrgOrThrow(undefined))) === null || _a === void 0 ? void 0 : _a.getUsername(); },
});
/**
 * An required org that is a devHub
 * Will throw if the specified org does not exist
 * Will default to the default dev hub if one is not specified
 * Will throw if no default deb hub exists and none is specified
 *
 * @example
 * import { SfCommand, Flags } from '@salesforce/sf-plugins-core';
 * public static flags = {
 *     // setting length or prefix
 *    'target-org': requiredHub(),
 *    // adding properties
 *    'flag2': requiredHub({
 *        required: true,
 *        description: 'flag2 description',
 *        char: 'h'
 *     }),
 * }
 */
exports.requiredHubFlag = core_1.Flags.build({
    char: 'v',
    parse: async (input) => await getHubOrThrow(input),
    default: async () => await getHubOrThrow(undefined),
    defaultHelp: async () => { var _a; return (_a = (await getHubOrThrow(undefined))) === null || _a === void 0 ? void 0 : _a.getUsername(); },
});
